MNT: replace __file__ with importlib.resources (compat with pyOxidizer and zipapp)

jackzyliu opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Closes #20081
helping #20150

#### What does this implement/fix? Explain your changes.
- Replaces occurrences of `__file__` with `importlib.resources` in `sklearn/datasets/*`
- Compatibility with `pyOxidizer` and `zipapp`: 
  - change IO functions to use `importlib.resources.{open, read}_{binary, text}`
  - a few path-based IO tests use `importlib.resources.path` to avoid making the assumption that resources already live in a filesystem

#### Any other comments?
- What is the best way to test compatibility? Something like [this](https://github.com/pytest-dev/pytest/issues/5780#issuecomment-524441969)? 

----
`sklearn/datasets`:
```
- [x] sklearn/datasets/_base.py:357:    module_path = dirname(__file__)
- [x] sklearn/datasets/_base.py:484:    module_path = dirname(__file__)
- [x] sklearn/datasets/_base.py:601:    module_path = dirname(__file__)
- [x] sklearn/datasets/_base.py:750:    module_path = dirname(__file__)
- [x] sklearn/datasets/_base.py:857:    module_path = dirname(__file__)
- [x] sklearn/datasets/_base.py:956:    base_dir = join(dirname(__file__), "data/")
- [x] sklearn/datasets/_base.py:970:    with open(dirname(__file__) + "/descr/linnerud.rst") as f:
- [x] sklearn/datasets/_base.py:1052:    module_path = dirname(__file__)
- [x] sklearn/datasets/_base.py:1122:    module_path = join(dirname(__file__), "images")
- [x] sklearn/datasets/_kddcup99.py:205:    module_path = dirname(__file__)
- [x] sklearn/datasets/_lfw.py:338:    module_path = dirname(__file__)
- [x] sklearn/datasets/_lfw.py:528:    module_path = dirname(__file__)
- [x] sklearn/datasets/_rcv1.py:281:    module_path = dirname(__file__)
- [x] sklearn/datasets/_california_housing.py:176:    module_path = dirname(__file__)
- [x] sklearn/datasets/_olivetti_faces.py:140:    module_path = dirname(__file__)
- [x] sklearn/datasets/tests/test_svmlight_format.py:19:currdir = os.path.dirname(os.path.abspath(__file__))
- [x] sklearn/datasets/tests/test_openml.py:36:currdir = os.path.dirname(os.path.abspath(__file__))
```

Other (should probably be separate PR(s) since `__file__` is used slightly differently in these occurrences):
```
- sklearn/__check_build/__init__.py:19:    local_dir = os.path.split(__file__)[0]
- sklearn/utils/__init__.py:1164:    root = str(Path(__file__).parent.parent)  # sklearn package
- sklearn/utils/_testing.py:733:        cwd = op.normpath(op.join(op.dirname(sklearn.__file__), ".."))
```

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

jackzyliu committed:
MNT: replace __file__ with importlib.resources (compat with pyOxidizer and zipapp)

sklearn/datasets/_base.py

jackzyliu commented on sklearn/datasets/_base.py:
```diff
@@ -22,6 +23,10 @@
 
+DATA_MODULE = "sklearn.datasets.data"
+DESCR_MODULE = "sklearn.datasets.descr"
+IMAGES_MODULE = "sklearn.datasets.images"
```
Both string and module work. Using string here so that it can be more conveniently serialized (e.g. `pickle`) as part of a `Bunch`.

jackzyliu commented on sklearn/datasets/_base.py:
```diff
@@ -659,7 +682,8 @@ def load_breast_cancer(*, return_X_y=False, as_frame=False):
         feature_names=feature_names,
-        filename=csv_filename,
+        filename=data_file_name,
+        file_module=DATA_MODULE
```
To be compatible with `pyOxidizer` and `zipapp`, we can no longer pass around an absolute file path. Instead, we separately pass around `filename` (just the name of the file) and `file_module` that the file lives in.

jackzyliu commented on sklearn/datasets/tests/test_base.py:
```diff
@@ -188,7 +189,11 @@ def test_loader(loader_func, data_shape, target_shape, n_target, has_descr, file
-        assert all([os.path.exists(bunch.get(f, False)) for f in filenames])
+        assert all([
+            f in bunch and 'file_module' in bunch and
+            importlib_resources.is_resource(bunch['file_module'], bunch[f])
```
Modifying the test here to remove checks on the absolute paths.

jackzyliu commented on sklearn/datasets/_base.py:
```diff
@@ -1119,16 +1143,17 @@ def load_sample_images():
+    descr = load_descr(IMAGES_MODULE, 'README.txt')
+
+    images_module_traversable = importlib_resources.files(IMAGES_MODULE)
+    with importlib_resources.as_file(images_module_traversable) as images_dir:
```
I think this is only available `python>=3.9`, so it could be problematic. So we probably need to find another (hopefully clean) way to do this.

jackzyliu replied:
We can use the backport library `importlib_resources` for `python<=3.8` as noted [here](https://importlib-resources.readthedocs.io/en/latest/).

quendee replied:
@jackzyliu Do you know if this is the reason for the failures in the PR? If so shall we try to make sure everything else is working and leave the images for after?

quendee replied:
@jackzyliu Do you know if this is the reason for the failures in the PR? If so shall we try to make sure everything else is working and leave the images for after?

jackzyliu replied:
Yea, it is. 

I will make the images portion backward-compatible.

jackzyliu committed:
Add resource modules as exceptions to test_all_tests_are_importable (GH20081)

ogrisel commented on sklearn/datasets/_base.py:
```diff
@@ -264,22 +269,47 @@ def load_data(module_path, data_file_name):
-            data[i] = np.asarray(ir[:-1], dtype=np.float64)
-            target[i] = np.asarray(ir[-1], dtype=int)
+    with importlib_resources.path(data_module, data_file_name) as data_path:
+        with open(data_path) as csv_file:
```
we should not `open(data_path)` assuming `data_path` is a valid filesystem path. Instead we should use [importlib.resource.open_binary](https://docs.python.org/3/library/importlib.html#importlib.resources.open_binary) or [importlib.resource.open_text](https://docs.python.org/3/library/importlib.html#importlib.resources.open_text) and pass the result to the `csv_reader`. I think for the `csv_reader` we need to open the resource in text mode, and probably hard-code `encoding="utf-8"` to avoid platform-dependent behaviors.

ogrisel commented on sklearn/datasets/_base.py:
```diff
@@ -264,22 +269,47 @@ def load_data(module_path, data_file_name):
+    string; content of descr_file_name
+    """
+
+    with importlib_resources.path(descr_module, descr_file_name) as data_path:
+        with open(data_path) as rst_file:
+            fdescr = rst_file.read()
```
This can probably be rewritten as:

```suggestion
    fdescr = importlib_resources.read_text(descr_module, descr_file_name)
```

to avoid assuming at any point that `descr_file_name` is the name of an actual file on a file system.

ogrisel reviewed (commented):
The point I a expressed in https://github.com/scikit-learn/scikit-learn/pull/20150#issuecomment-863393743 is to avoid use the `open()` Python builtin anywhere and instead using the functions from `importlib.resources.open_binary/text` or `importlib.resources.read_binary/text` documented here:

https://docs.python.org/3/library/importlib.html#module-importlib.resources

The goal is to avoid assuming those resource names are actual filenames of resource stored on a filesystem at any point.

jackzyliu commented:
> The goal is to avoid assuming those resource names are actual filenames of resource stored on a filesystem at any point.

Gotcha. I will make these changes accordingly and make the `.files` portion backward-compatible.

jackzyliu commented on sklearn/datasets/_base.py:
```diff
@@ -264,7 +269,7 @@ def load_data(module_path, data_file_name):
         target_names[0] is the name of the target[0] class.
     """
-    with open(join(module_path, "data", data_file_name)) as csv_file:
+    with importlib_resources.open_text(data_module, data_file_name) as csv_file:
```
`.open_text` has `encoding="utf-8"` by default ([reference](https://github.com/python/cpython/blame/3.9/Lib/importlib/resources.py#L115-L118))

Happy to make it explicit though.

ogrisel replied:
I think this is fine to rely on the default utf-8 encoding assumption. I am pretty sure that all our CSV file only need the ascii encoding but this is a valid subset of utf-8 as far as I know.

We just needed to ensure that the loader would not implicitly try to use a platform dependent encoding.

jackzyliu commented on sklearn/datasets/_base.py:
```diff
@@ -280,6 +285,67 @@ def load_data(module_path, data_file_name):
     return data, target, target_names
 
 
+def load_gzip_compressed_csv_data(data_module, data_file_name, **kwargs):
```
I don't think we can ask `np.loadtxt` to decompress if we are passing in file object (as opposed to a file name ending with '.gz'). So for the few `.csv.gz` files, we first manually decompress the file object and the use `np.loadtxt` to load data as before.

jackzyliu commented on sklearn/datasets/tests/test_base.py:
```diff
@@ -122,6 +128,39 @@ def test_load_files_wo_load_content(
+        ("breast_cancer.csv", 569, 30, ["malignant", "benign"]),
+    ],
+)
+def test_load_csv_data(
```
Adding tests for `load_csv_data` and `load_gzip_compressed_csv_data`.

jackzyliu commented on sklearn/datasets/_base.py:
```diff
@@ -1119,16 +1175,16 @@ def load_sample_images():
+
+    filenames = list()
+    images = list()
+    for filename in sorted(importlib_resources.contents(IMAGES_MODULE)):
```
We can directly use `.contents()` (which returns a list of file names and is available since `python 3.7`) to avoid using `.files()`. It was my mistake not remembering the option.

jackzyliu commented:
@ogrisel

Just updated according to your comments. Is it more along the line of what you were thinking?

I think the failing tests come from main. Will rebase after those are fixed.

jackzyliu committed:
replace __file__ with importlib.resources for sklearn/datasets/*.py (GH20081)

jackzyliu committed:
replace __file__with importlib.resources for test_openml.py (GH20081)

jackzyliu committed:
replace __file__ with importlib.resources for test_svmlight_format.py (GH20081)

jackzyliu commented:
@ogrisel @quendee 

At this point, all the conversions are done for `sklearn/datasets`. I think we should probably have separate PR(s) for the other 3 occurrences (see below: e.g. `sklearn/__check_build/__init__.py:19`) since `__file__` is used slightly differently in these cases.

For ease of review, I have a separate commit for each of the following component:
1. `sklearn/datasets/_base.py`
2. test adjustment for `test_common.py`
3. `sklearn/datasets/{_californa_housing, _covtype, _kddcup99, ...}.py`: just the same changes applied to `_base.py`
4. `sklearn/datasets/tests/test_openml.py`: This involves changing data directory names from digits to `id_*` to make them valid module names (thus the large number of file changes), but otherwise straightforward.
5. `sklearn/datasets/tests/test_svmlight_format.py`: A few tests explicitly test path-based input for `load_svmlight_file` and `load_svmlight_files`. So for those tests, I used `importlib.resources.path` (to generate a temporary path if necessary), but otherwise used `importlib.resources.open_binary` as much as possible. Hope this is ok.

Let me know of comments. 

----

`sklearn/datasets`:
```
- [x] sklearn/datasets/_base.py:357:    module_path = dirname(__file__)
- [x] sklearn/datasets/_base.py:484:    module_path = dirname(__file__)
- [x] sklearn/datasets/_base.py:601:    module_path = dirname(__file__)
- [x] sklearn/datasets/_base.py:750:    module_path = dirname(__file__)
- [x] sklearn/datasets/_base.py:857:    module_path = dirname(__file__)
- [x] sklearn/datasets/_base.py:956:    base_dir = join(dirname(__file__), "data/")
- [x] sklearn/datasets/_base.py:970:    with open(dirname(__file__) + "/descr/linnerud.rst") as f:
- [x] sklearn/datasets/_base.py:1052:    module_path = dirname(__file__)
- [x] sklearn/datasets/_base.py:1122:    module_path = join(dirname(__file__), "images")
- [x] sklearn/datasets/_kddcup99.py:205:    module_path = dirname(__file__)
- [x] sklearn/datasets/_lfw.py:338:    module_path = dirname(__file__)
- [x] sklearn/datasets/_lfw.py:528:    module_path = dirname(__file__)
- [x] sklearn/datasets/_rcv1.py:281:    module_path = dirname(__file__)
- [x] sklearn/datasets/_california_housing.py:176:    module_path = dirname(__file__)
- [x] sklearn/datasets/_olivetti_faces.py:140:    module_path = dirname(__file__)
- [x] sklearn/datasets/tests/test_svmlight_format.py:19:currdir = os.path.dirname(os.path.abspath(__file__))
- [x] sklearn/datasets/tests/test_openml.py:36:currdir = os.path.dirname(os.path.abspath(__file__))
```

Others (should probably be separate PR(s) since `__file__` is used slightly differently in these occurrences):
```
- sklearn/__check_build/__init__.py:19:    local_dir = os.path.split(__file__)[0]
- sklearn/utils/__init__.py:1164:    root = str(Path(__file__).parent.parent)  # sklearn package
- sklearn/utils/_testing.py:733:        cwd = op.normpath(op.join(op.dirname(sklearn.__file__), ".."))
```

jackzyliu committed:
DOC add whats new (GH20081)

jackzyliu committed:
TST add descr checks for data fetch functions in sklearn/datasets (GH20081)

jackzyliu commented:
Updated whats new, and added a few `.DESCR` checks for the few modified data fetch functions in `sklearn/datasets` (e.g. `fetch_olivetti_faces`).

ogrisel commented on sklearn/datasets/_base.py:
```diff
@@ -280,6 +285,67 @@ def load_data(module_path, data_file_name):
+            gzip_encoding = kwargs["encoding"]
+
+        compressed_file = gzip.open(compressed_file, mode="rt", encoding=gzip_encoding)
+        data = np.loadtxt(compressed_file, **kwargs)
```
I don't understand why we would pass the `encoding` kwarg to `np.loadtxt` if it has arleady been used to decode the text of `gzip.open`.

Either call `gzip.open` with `mode="rb"` and only pass encoding to `np.loadtxt` so that it handles the text decoding correctly itself or pass it only to `gzip.open` with `model="rt"` (in which case it would make sense to have `load_gzip_compressed_csv_data(data_module, data_file_name, encoding="utf-8", **kwargs)` to simplify the code to only forward `kwargs` to `np.loadtxt`.

ogrisel commented on sklearn/datasets/_base.py:
```diff
@@ -280,6 +285,67 @@ def load_data(module_path, data_file_name):
+    -------
+    string; content of descr_file_name
+    """
+    fdescr = importlib_resources.read_text(descr_module, descr_file_name)
+
+    return fdescr
```
As far as I can see most of the descriptions are loaded from `DESCR_MODULE` so to avoid redundancy we could have:

```
load_descr(descr_file_name, descr_module=DESCR_MODULE):
```

that would make it possible to only import `load_descr` without having to import the `DESCR_MODULE` constant most of the time.

ogrisel replied:
For convenience you could add an optional `descr_file_name=None` / `descr_module=DESCR_MODULE` to `load_csv_data` and `load_gzip_compressed_csv_data` and when `descr_file_name` is not `None`, make those functions return the loaded description as as additional return value.

jackzyliu replied:
> As far as I can see most of the descriptions are loaded from `DESCR_MODULE` so to avoid redundancy we could have:
> 
> ```
> load_descr(descr_file_name, descr_module=DESCR_MODULE):
> ```
> 
> that would make it possible to only import `load_descr` without having to import the `DESCR_MODULE` constant most of the time.

By extension, I am applying the same changes to `load_csv_data` and `load_gzip_compressed_csv_data` to avoid the same redundancy since these data are always loaded from `DATA_MODULE`.  
```
load_csv_data(data_file_name, data_module=DATA_MODULE)
```
And since as positional arguments to `importlib.resources` functions, the resource module always comes before the resource name, I am enforcing named arguments to avoid usage error.
```
load_csv_data(data_file_name, *, data_module=DATA_MODULE)
load_descr(descr_file_name, *, descr_module=DESCR_MODULE)
```
Lastly, combined with the suggestion to add optional `descr_file_name=None, descr_module=DESCR_MODULE`, we have the following function signature:
```
load_csv_data(data_file_name, *, data_module=DATA_MODULE, descr_file_name=None, descr_module=DESCR_MODULE)
```

ogrisel commented on sklearn/datasets/_base.py:
```diff
@@ -280,6 +285,67 @@ def load_data(module_path, data_file_name):
+            gzip_encoding = kwargs["encoding"]
+
+        compressed_file = gzip.open(compressed_file, mode="rt", encoding=gzip_encoding)
+        data = np.loadtxt(compressed_file, **kwargs)
```
I don't understand why we would pass the `encoding` kwarg to `np.loadtxt` if it has arleady been used to decode the text of `gzip.open`.

Either call `gzip.open` with `mode="rb"` and only pass encoding to `np.loadtxt` so that it handles the text decoding correctly itself or pass it only to `gzip.open` with `model="rt"` (in which case it would make sense to have `load_gzip_compressed_csv_data(data_module, data_file_name, encoding="utf-8", **kwargs)` to simplify the code to only forward `kwargs` to `np.loadtxt`.

jackzyliu replied:
Good point. Although `encoding` is passed around a bunch of times in `np.loadtxt`, it's not really used when the file is already in text.

Going with the latter option.

ogrisel commented on sklearn/datasets/_base.py:
```diff
@@ -481,12 +545,9 @@ def load_iris(*, return_X_y=False, as_frame=False):
-
-    with open(join(module_path, "descr", "iris.rst")) as rst_file:
-        fdescr = rst_file.read()
+    data_file_name = "iris.csv"
+    data, target, target_names = load_csv_data(DATA_MODULE, data_file_name)
```
I think we can avoid the extra indirection in a local variable that is used only once.

```suggestion
    data, target, target_names = load_csv_data(DATA_MODULE, "iris.csv")
```

jackzyliu replied:
`data_file_name` is actually used again in constructing the `Bunch` below, which is why I used a variable. Do you still want me to make the change?

ogrisel commented on sklearn/datasets/_base.py:
```diff
@@ -598,12 +660,9 @@ def load_breast_cancer(*, return_X_y=False, as_frame=False):
-
-    with open(join(module_path, "descr", "breast_cancer.rst")) as rst_file:
-        fdescr = rst_file.read()
+    data_file_name = "breast_cancer.csv"
+    data, target, target_names = load_csv_data(DATA_MODULE, data_file_name)
```
```suggestion
    data, target, target_names = load_csv_data(DATA_MODULE, "breast_cancer.csv")
```

ogrisel commented on sklearn/datasets/tests/test_base.py:
```diff
@@ -188,7 +227,14 @@ def test_loader(loader_func, data_shape, target_shape, n_target, has_descr, file
     if has_descr:
         assert bunch.DESCR
     if filenames:
```
If we expect a non-empty list of filenames, I think we should require the data module to be present in the bunch.

ogrisel commented on sklearn/datasets/tests/test_base.py:
```diff
@@ -188,7 +227,14 @@ def test_loader(loader_func, data_shape, target_shape, n_target, has_descr, file
+        assert all(
+            [
+                f in bunch
+                and "file_module" in bunch
```
I would rather use the `data_module` field in all those Bunch instances:

```suggestion
                and "data_module" in bunch
```

ogrisel commented on sklearn/datasets/tests/test_openml.py:
```diff
@@ -1445,14 +1441,17 @@ def test_fetch_openml_verify_checksum(monkeypatch, as_frame, cache, tmpdir):
-    ) as modified_gzip:
+    original_data_module = OPENML_TEST_DATA_MODULE + "." + f"id_{data_id}"
+    original_data_file_name = "data-v1-dl-1666876.arff.gz"
+    corrupt_copy_path = os.path.join(tmpdir, "test_invalid_checksum.arff")
```
nitpick: I am pretty sure this can be simplified as:

```suggestion
    corrupt_copy_path = tmpdir / "test_invalid_checksum.arff"
```

ogrisel commented on sklearn/datasets/tests/test_svmlight_format.py:
```diff
@@ -5,6 +5,7 @@
 import scipy.sparse as sp
 import os
 import shutil
+from importlib import resources as importlib_resources
```
I think we could just use:

```suggestion
from importlib import resources
```

and then call `resources.open_binary` / `open_binary.open_text` instead of renaming this module every where.

ogrisel reviewed (commented):
Thanks @jackzyliu. This look good overall. Here are a few suggestion for further improvement. Once done I think that would be ready for merge on my end.

jackzyliu committed:
Updates according to comments (GH20081)

- change signatures of underlying load functions to reduce various redundancies
- allow `load_descr` within `load_csv_data` and `load_gzip_compressed_csv_data`
- undo renaming of `importlib.resources` in import
- rename module field in returning `Bunch`
- other style improvements

jackzyliu commented:
`codecov/patch` shows failed, but I think the marked statements are all either 1) tested in 9cd899a0fab84a41b7a3ad858d62ab36a5e720f3 (descr checks) or 2) themselves test statements.

glemaitre commented:
Regarding the failure, I don't think this is big deal. We need to be sure to run the test locally (they are passing on my side) but they will be skipped anyway on the CIs.

glemaitre commented on sklearn/datasets/_base.py:
```diff
@@ -238,17 +243,34 @@ def load_files(
     data_file_name : string
         Name of csv file to be loaded from
-        module_path/data/data_file_name. For example 'wine_data.csv'.
+        data_module/data_file_name. For example 'wine_data.csv'.
```
```suggestion
        `data_module/data_file_name`. For example `'wine_data.csv'`.
```

glemaitre commented on sklearn/datasets/_base.py:
```diff
@@ -238,17 +243,34 @@ def load_files(
-        module_path/data/data_file_name. For example 'wine_data.csv'.
+        data_module/data_file_name. For example 'wine_data.csv'.
+
+    data_module : string or module, optional
+        module where data lives;
+        default "sklearn.datasets.data" (DATA_MODULE).
```
```suggestion
    data_module : str or module, default="sklearn.datasets.data"
        Module where data lives. The default is `"sklearn.datasets.data"`.
```

glemaitre commented on sklearn/datasets/_base.py:
```diff
@@ -238,17 +243,34 @@ def load_files(
+        module where data lives;
+        default "sklearn.datasets.data" (DATA_MODULE).
+
+    descr_file_name : string, optional
```
```suggestion
    descr_file_name : str, default=None
```

glemaitre commented on sklearn/datasets/_base.py:
```diff
@@ -238,17 +243,34 @@ def load_files(
+        default "sklearn.datasets.data" (DATA_MODULE).
+
+    descr_file_name : string, optional
+        (See `load_descr`) Name of rst file to be loaded from
+        descr_module/descr_file_name. For example 'wine_data.rst';
+        If not None, also returns the corresponding description of
+        the dataset.
```
```suggestion
        Name of rst file to be loaded from `descr_module/descr_file_name`.
        For example 'wine_data.rst'. See also :func:`load_descr`. 
        If not None, also returns the corresponding description of
        the dataset.
```

glemaitre commented on sklearn/datasets/_base.py:
```diff
@@ -238,17 +243,34 @@ def load_files(
+        If not None, also returns the corresponding description of
+        the dataset.
+
+    descr_module : string or module, optional
```
```suggestion
    descr_module : str or module, default="sklearn.datasets.descr"
```

glemaitre commented on sklearn/datasets/_base.py:
```diff
@@ -238,17 +243,34 @@ def load_files(
+        the dataset.
+
+    descr_module : string or module, optional
+        (See `load_descr`) module where descr_file_name lives;
+        default "sklearn.datasets.descr" (DESCR_MODULE);
```
```suggestion
        Module where `descr_file_name` lives. See also :func:`load_descr`.
        The default  is `"sklearn.datasets.descr"`.
```

glemaitre commented on sklearn/datasets/_base.py:
```diff
@@ -263,8 +285,12 @@ def load_data(module_path, data_file_name):
     target_names : Numpy array
```
```suggestion
    target_names : ndarray
```

glemaitre commented on sklearn/datasets/_base.py:
```diff
@@ -263,8 +285,12 @@ def load_data(module_path, data_file_name):
         A 1D array containing the names of the classifications. For example
         target_names[0] is the name of the target[0] class.
+
+    descr : string, optional
```
```suggestion
    descr : str, optional
```

glemaitre commented on sklearn/datasets/_base.py:
```diff
@@ -263,8 +285,12 @@ def load_data(module_path, data_file_name):
         target_names[0] is the name of the target[0] class.
+
+    descr : string, optional
+        Description of the dataset (content of descr_file_name). Only returned
+        if descr_file_name is not None.
```
```suggestion
        Description of the dataset (the content of `descr_file_name`). Only returned
        if `descr_file_name` is not None.
```

glemaitre commented on sklearn/datasets/_base.py:
```diff
@@ -277,7 +303,99 @@ def load_data(module_path, data_file_name):
+    encoding="utf-8",
+    **kwargs,
+):
+    """Loads data_file_name from data_module with importlib.resources and np.loadtxt
```
```suggestion
    """Load `data_file_name` from `data_module` with `importlib.resources` and `np.loadtxt`.
```

glemaitre commented on sklearn/datasets/_base.py:
```diff
@@ -277,7 +303,99 @@ def load_data(module_path, data_file_name):
+
+    Parameters
+    ----------
+    data_file_name : string
```
```suggestion
    data_file_name : str
```

glemaitre commented on sklearn/datasets/_base.py:
```diff
@@ -277,7 +303,99 @@ def load_data(module_path, data_file_name):
+    ----------
+    data_file_name : string
+        Name of gzip-compressed csv (.csv.gz) file to be loaded from
+        data_module/data_file_name. For example 'diabetes_data.csv.gz'.
```
```suggestion
        `data_module/data_file_name`. For example `'diabetes_data.csv.gz'`.
```

glemaitre commented on sklearn/datasets/_base.py:
```diff
@@ -277,7 +303,99 @@ def load_data(module_path, data_file_name):
+        Name of gzip-compressed csv (.csv.gz) file to be loaded from
+        data_module/data_file_name. For example 'diabetes_data.csv.gz'.
+
+    data_module : string or module, optional
+        module where data lives;
+        default "sklearn.datasets.data" (DATA_MODULE).
+
+    descr_file_name : string, optional
+        (See `load_descr`) Name of rst file to be loaded from
+        descr_module/descr_file_name. For example 'wine_data.rst';
+        If not None, also returns the corresponding description of
+        the dataset.
+
+    descr_module : string or module, optional
+        (See `load_descr`) module where descr_file_name lives;
+        default "sklearn.datasets.descr" (DESCR_MODULE);
```
Could you format those as suggested in the previous function

glemaitre commented on sklearn/datasets/_base.py:
```diff
@@ -277,7 +303,99 @@ def load_data(module_path, data_file_name):
+        (See `load_descr`) module where descr_file_name lives;
+        default "sklearn.datasets.descr" (DESCR_MODULE);
+
+    encoding : string, optional
+        Name of the encoding that the gzip-decompressed file will be
+        decoded with; default "utf-8"
```
```suggestion
    encoding : str, default="utf-8"
        Name of the encoding that the gzip-decompressed file will be
        decoded with. The default is "utf-8".
```

glemaitre commented on sklearn/datasets/_base.py:
```diff
@@ -277,7 +303,99 @@ def load_data(module_path, data_file_name):
+
+    **kwargs : dict, optional
+        Keyword arguments to be passed to `np.loadtxt`;
+        e.g. delimiter=","
```
```suggestion
        e.g. delimiter=",".
```

glemaitre commented on sklearn/datasets/_base.py:
```diff
@@ -277,7 +303,99 @@ def load_data(module_path, data_file_name):
+
+    Returns
+    -------
+    data : Numpy array
+        A 2D array with each row representing one sample and each column
+            representing the features and/or target of a given sample.
```
```suggestion
    data : ndarray of shape (n_samples, n_features)
        A 2D array with each row representing one sample and each column
        representing the features and/or target of a given sample.
```

glemaitre commented on sklearn/datasets/_base.py:
```diff
@@ -277,7 +303,99 @@ def load_data(module_path, data_file_name):
+        A 2D array with each row representing one sample and each column
+            representing the features and/or target of a given sample.
+
+    descr : string, optional
+        Description of the dataset (content of descr_file_name). Only returned
+        if descr_file_name is not None.
```
Can you format it as in the previous function.

glemaitre commented on sklearn/datasets/_base.py:
```diff
@@ -277,7 +303,99 @@ def load_data(module_path, data_file_name):
+
+    Parameters
+    ----------
+    descr_file_name : string
+        Name of rst file to be loaded from
+        descr_module/descr_file_name. For example 'wine_data.rst'.
+
+    descr_module : string or module, optional
+        module where descr_file_name lives;
+        default "sklearn.datasets.descr" (DESCR_MODULE).
```
If you use the docstring of the previous function as well.

glemaitre commented on sklearn/datasets/_base.py:
```diff
@@ -277,7 +303,99 @@ def load_data(module_path, data_file_name):
+
+
+def load_descr(descr_file_name, *, descr_module=DESCR_MODULE):
+    """Loads descr_file_name from descr_module with importlib.resources.
```
```suggestion
    """Load `descr_file_name` from `descr_module` with `importlib.resources`.
```

glemaitre commented on sklearn/datasets/_base.py:
```diff
@@ -277,7 +303,99 @@ def load_data(module_path, data_file_name):
+
+    Returns
+    -------
+    string; content of descr_file_name
```
```suggestion
    fdescr : str
        Content of `descr_file_name`.
```

glemaitre commented on sklearn/datasets/_base.py:
```diff
@@ -1119,16 +1231,16 @@ def load_sample_images():
-    images = [imread(filename) for filename in filenames]
+    descr = load_descr("README.txt", descr_module=IMAGES_MODULE)
+
+    filenames = list()
+    images = list()
```
```suggestion
    filenames, images = [], []
```

glemaitre reviewed (approved):
LGTM. This only suggestions for the documentation to follow as much as possible numpydoc style

jackzyliu committed:
Modify function documentation to better follow Numpy docstring style (GH20081)

jackzyliu commented:
@glemaitre Thanks a lot for the review and suggestions! I have modified the documentations accordingly. Let me know if anything else needs changing.

Weirdly, the doc checks both seem have failed at `build_doc.sh` and specifically `pip install "$(get_dep sphinxext-opengraph $SPHINXEXT_OPENGRAPH_VERSION)"` where `$SPHINXEXT_OPENGRAPH_VERSION` didn't read in the values specified in `config.yml`. 

I will wait for further comments before I push to force-run checks again.

glemaitre commented:
I think that you only need to merge `main` into your branch and it should be fine.

jackzyliu commented:
> I think that you only need to merge `main` into your branch and it should be fine.

Thanks! All green now.

rth reviewed (approved):
Thanks @jackzyliu ! LGTM as well. Merging with two approvals above.

nithish08 reviewed (commented):
LGTM
